Update C++ rules (#3220)
SonarSource/rspec: rules/S6187/cfamily/rule.adoc

== Why is this an issue?

{cpp}20 introduces the "spaceship" ``++operator<=>++`` that replaces all the other comparison operators in most cases. When this operator is defined, the compiler can rewrite expressions using ``++<++``, ``++<=++``, ``++>++`` and ``++>=++`` to use this operator instead. This presents three advantages:

* Less code to write (and therefore fewer bugs, too),
* Guaranteed consistency between all the comparison operators (for instance, in this situation, ``++a < b++`` and ``++!(a >= b)++`` will always return the same value).

> * Guaranteed symmetry for comparisons: If you can write ``++a<b++``, and that operation is resolved through ``++operator<=>++``, you can also write ``++b<a++``, and get a consistent result. Achieving the same result with classical comparison operators requires double the number overloads if ``++a++`` and ``++b++`` have different types.

Reviewer: Small nit on spaces around operator:
```suggestion
* Guaranteed symmetry for comparisons: If you can write ``++a < b++``, and that operation is resolved through ``++operator<=>++``, you can also write ``++b < a++``, and get a consistent result. Achieving the same result with classical comparison operators requires double the number overloads if ``++a++`` and ``++b++`` have different types.
```

Another suggestion (Grammarly complained a bit about the current wording):
```suggestion
* Guaranteed symmetry for comparisons: If you can write ``++a < b++``, and that operation is resolved through ``++operator<=>++``, you can also write ``++b < a++``, and get a consistent result. Achieving the same result with classical comparison operators requires twice as many overloads if ``++a++`` and ``++b++`` have different types.
```

> Additionally, if the ``++operator<=>++`` has the defaulted implementation, the compiler can implicitly generate a defaulted implementation of operator==, simplifying the class definition one step further.

Reviewer: ```suggestion
Additionally, if the ``++operator<=>++`` has the defaulted implementation, the compiler can implicitly generate a defaulted implementation of `operator==`, simplifying the class definition one step further.
```

> Before {cpp}20, it was common to provide only ``++operator<++`` for a class and ask the user of this class to write all his code only using this operator (this is what ``++std::map++`` requires of its key type, for instance). In this case, it is still advised to replace the operator with ``++<=>++``: the quantity of required work is similar, and the class user will benefit from a much greater expressivity.

Reviewer: ```suggestion
Before {cpp}20, it was common to provide only ``++operator<++`` for a class and ask the user of this class to write all their code only using this operator (this is what ``++std::map++`` requires of its key type, for instance). In this case, it is still advised to replace the operator with ``++<=>++``: the quantity of required work is similar, and users of the class will benefit from a much greater expressivity.
```

== How to fix it

> This rule reports user-provided comparison operators (member functions or free functions) ``++<++``, ``++<=++``, ``++>++`` and ``++>=++`` that could be replaced by defining the ``++operator<=>++`.

Reviewer: A ` is missing:
```suggestion
This rule reports user-provided comparison operators (member functions or free functions) ``++<++``, ``++<=++``, ``++>++`` and ``++>=++`` that could be replaced by defining the ``++operator<=>++``.
```

== Resources

=== Documentation

* {cpp} reference - https://en.cppreference.com/w/cpp/language/operator_comparison#Three-way_comparison[Three-way comparison]

=== Related rules

> * S6186 - Redundant comparison operators should not be defined.

Reviewer: Should we also link to S6230?

Author: Yes, added in https://github.com/SonarSource/rspec/pull/3220/commits/cf9d919c18a7db9da20a9d8d1b688aca80e25aa9
